packages/react-reconciler/src/ReactMemoCache.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {MemoCache} from 'shared/ReactTypes';
+import type {Dispatcher} from './ReactInternalTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentDispatcher} = ReactSharedInternals;
+
+export const MEMO_CACHE_UNSET_SENTINEL: symbol = Symbol.for(
+  'react.usememocache_sentinel',
+);
+
+const MemoCachePrototype: MemoCache = {
+  has(index: number): boolean {
+    // $FlowFixMe[object-this-reference]
+    return this.data[index] !== MEMO_CACHE_UNSET_SENTINEL;
+  },
+
+  get(index: number): any {
+    // $FlowFixMe[object-this-reference]
+    return this.data[index];
+  },
+
+  update(index: number, value: any): boolean {
+    // $FlowFixMe[object-this-reference]
+    const previous = this.data[index];
+    // $FlowFixMe[object-this-reference]
+    this.data[index] = value;
+    return previous === MEMO_CACHE_UNSET_SENTINEL || previous !== value;
+  },
+
+  withDispatcher_TEMPORARY_DEBUG_API_DO_NOT_USE<T>(
+    dispatcher: Dispatcher,
+    fn: () => T,
+  ): T | void {
+    const previousDispatcher = ReactCurrentDispatcher.current;
+    ReactCurrentDispatcher.current = dispatcher;
+    try {
+      return fn();
+    } finally {
+      ReactCurrentDispatcher.current = previousDispatcher;
+    }
+  },
+};
+
+export function createMemoCache(data: Array<any>): MemoCache {
+  const cache: {data: Array<any>} = (Object.create(MemoCachePrototype): any);
+  cache.data = data;
```

Comment:
not sure how much we care about perf/memory here, but one trick is to just use the `cache` object as an array and index into it directly rather than creating a separate `data` array